### PR TITLE
release/0.10.2

### PR DESCRIPTION
**Pull request by f-wllr**
### New app features:

- add support for iOS bluetooth connections
- add configurable vibration alerts for timers
- add phone notifications for configured timers if app is in background
- add ability to change robots label
- add ability to keep the screen awake
- add ability to "save" individual modules
- add ability to "save" presets for multiple modules

---

### Bugfixes:

- fix timers to "continue" running in the background

---

### OTHER:

**IMPORTANT, PLEASE READ**

I Introduced a new git workflow, including new CI/CD pipelines that automatically provide debug builds and releases.

General workflow:

Branches: 

- `main` --> stable, contains releases
- `dev` --> debug, contains new features
- `feat/*` --> individual feature development branches, contain just one new feature

This also includes `fix/*` branches, that are just used to fix one thing, but as everything works the first try, we don't need them xD

As described above, the `main` branch only contains releases and it is prohibited to push to `main` via rulesets (needs to be added manually). If a push (merge) to `main` happens, the new version is extracted from pubspec.yaml, a git tag is created with that version, an android release app is build and a GitHub release is created with automatic change notes and the android apk.

The `dev` branch is always co-existing with the `main` branch and collects features prior to a release. For each push (merge) to the `dev` branch a debug apk is build and provided as an artifact.

All feature branches (`feat/*`) are used to work on one feature and if done, should be merged in the `dev` branch. For each push the app is build to ensure runnable code.

Also, for each PR there is a "quality gate" that runs `flutter analyze` and `flutter test` and if one of this fails the code is classified as "needs improvement (testing is planned for the future and currently passes as nothing is actually tested).

### CODE REVIEW

**Comment by f-wllr** [comment]
Before this can be merged we need to create the keys for signing the release apk and storing the values in the repos secrets

**Comment by mato157** [comment]
For Android, Google Play App Signing is enabled, so the final release builds are signed by Google Play.

In the repository/CI secrets we only need to add the Android upload signing key/keystore. This key is used only to sign the AAB/APK before uploading it to Google Play, so Google can verify that the upload comes from us.

**Comment by mrshu** [comment]
@f-wllr I guess we are not reviewing this though, right? It should have all been reviewed before.

**Comment by f-wllr** [comment]
> @f-wllr I guess we are not reviewing this though, right? It should have all been reviewed before.

@mrshu Well, you said that we should also review @mato157 's commits, and this is the PR that also contains them

**Comment by mrshu** [comment]
@f-wllr ok, then let me run my battery here and let's try to close the other PR then

File: lib/models/game.dart [hunk 1]
```diff
@@ -238,22 +279,149 @@ class Game with ChangeNotifier {
...
+        // never add time back (clamp floor 0) and never auto-start the
+        // referee-controlled second half (_maxResumeCatchUpTicks halts at the
+        // half-time -> second-half manual gate, reinforced by the isTimeRunning
+        // guard below since that transition sets isTimeRunning = false).
+        final ticksToProcess = (elapsedSeconds - alreadyApplied)
+            .clamp(0, _maxResumeCatchUpTicks())
+            .toInt();
+        for (var i = 0; i < ticksToProcess && isTimeRunning; i++) {
+          _tickTimer();
+        }
```
Comment (mrshu) [inline lib/models/game.dart:403]: **[high] background-timer** — On resume this catch-up loop replays missed ticks synchronously, and `_maxResumeCatchUpTicks()` lets it cross the whole first-half + half-time window in one go.

Why: when first-half hits zero, `_tickTimer` calls `halfTimeAll()`, which does `stopAll(true)` then `await Future.delayed(1s)` before `module.halfTime()`. This loop has no awaits, so it reaches `secondHalf` (`stopAll(force:true)`) before that 1s elapses — the deferred `module.halfTime()` then fires against modules the game already treats as stopped/ready, dropping the robots back into a half-time countdown after the break is over.

Suggested fix: guard the deferred side effect on stage still being current — in `halfTimeAll()`, after the `await`, skip the `module.halfTime()` fan-out unless `currentStage == MatchStage.halfTime`; apply the same stale-stage guard to `gameOverAll()`. Lock it in with a lifecycle test: short period/halftime, pause past first-half + halftime, resume, wait > 1s, assert game ends in `secondHalf` with modules stopped.

File: lib/screens/settings.dart [hunk 2]
```diff
@@ -449,10 +448,82 @@ class _SettingsScreenState extends State<SettingsScreen> {
...
+                                SettingSwitch(
+                                  title: 'Game Timer Vibration',
+                                  value: vs.gameTimerEnabled,
+                                  onChanged: (value) {
+                                    vs.gameTimerEnabled = value;
+                                    // Ask for notification permission only now,
+                                    // when the user opts into timer alerts.
+                                    if (value) {
+                                      NotificationService.requestPermission();
+                                    }
```
Comment (mrshu) [inline lib/screens/settings.dart:469]: **[high] notifications** — Notification permission is requested only here, on an off→on toggle, but `VibrationService` defaults both alert flags to `true`, so a default install is never prompted.

Why: `NotificationService.initialize()` deliberately does not request permission, and the only `requestPermission()` calls are these two `onChanged` handlers (gated on `if (value)`). A fresh user who keeps the default-on switches never sees the OS dialog; when backgrounded mid-timer, `_scheduleBackgroundNotifications()` still calls `zonedSchedule()`, but Android 13+ `POST_NOTIFICATIONS` / iOS grant was never obtained, so every alert is dropped — the headline background-notification feature silently does nothing out of the box.

Suggested fix: request permission for the default-enabled state too — from `main()` after `initialize()` (or first foreground) when either alert is enabled, or before the first `_scheduleBackgroundNotifications()` schedule. Keep the lazy on-toggle prompt, guard with a one-shot flag, and add a test from empty prefs asserting `requestPermission` runs before a default-settings backgrounded timer schedules.

File: lib/models/game.dart [hunk 3]
```diff
@@ -238,22 +279,149 @@ class Game with ChangeNotifier {
...
+  }
+
+  void _scheduleBackgroundNotifications() {
+    // Game timer
+    if (vibrationService.gameTimerEnabled) {
+      switch (currentStage) {
+        case MatchStage.firstHalf:
+          NotificationService.scheduleGameAlerts(
+              _remainingTime, vibrationService.gameTimerAlerts);
+          break;
```
Comment (mrshu) [inline lib/models/game.dart:434]: **[medium] notifications** — Pausing in the first half schedules only the first-half game alerts, never the half-time break alerts.

Why: `_maxResumeCatchUpTicks()` treats first-half + half-time as one auto-run window, so if the app stays backgrounded past first-half expiry the break is considered "run" on resume — but no break notification was scheduled. The referee gets a "first half over" alert and then nothing for "half-time over / start second half" while away.

Suggested fix: in this `firstHalf` case, also schedule the break alerts offset by the remaining first-half duration. This requires distinct notification IDs — `scheduleGameAlerts` and `scheduleBreakAlerts` both use `10000 + threshold` (notification_service.dart L84 & L112), so co-scheduled alerts would collide; give break alerts their own ID base. Cover it with a fake-scheduler test (`remainingTime=60`, `halfTimeDuration=30`).

File: test/widget_test.dart [hunk 4]
```diff
@@ -1,25 +1,32 @@
...
+    // Tap the '+' icon and trigger a frame.
+    // await tester.tap(find.byIcon(Icons.add));
+    // await tester.pump();
 
-    expect(find.byType(MaterialApp), findsOneWidget);
-    expect(find.byType(Home), findsOneWidget);
+    // Verify that our counter has incremented.
+    // expect(find.text('0'), findsNothing);
+    // expect(find.text('1'), findsOneWidget);
   });
```
Comment (mrshu) [inline test/widget_test.dart:28]: **[medium] test-coverage** — The app smoke test is now an empty no-op (whole body commented out) in the same PR that adds new startup paths.

Why: `Game()` now registers a `WidgetsBindingObserver` and builds four new services, and `NotificationService.initialize()` runs at startup, yet the only widget test no longer builds `MyApp`. The new `pr-quality-gate.yml` runs `flutter test`, which now passes trivially and gives false confidence that startup is verified. (The PR description does acknowledge tests are deferred, so this is partly a known trade-off.)

Suggested fix: restore a real smoke test — `SharedPreferences.setMockInitialValues({})`, construct `Game()`, `pumpWidget(MyApp(game: game))`, assert `find.byType(Home)` — so the gate actually guards startup.

File: lib/models/game.dart [hunk 5]
```diff
@@ -238,22 +279,149 @@ class Game with ChangeNotifier {
...
+    if (vibrationService.damageTimerEnabled) {
+      for (final team in teams) {
+        for (final module in team.modules.where(
+            (m) => m.isEnabled && m.state == ModuleState.damage && m.penaltyTime > 0)) {
+          NotificationService.scheduleDamageAlerts(
+              module.moduleId, module.name, module.penaltyTime,
+              vibrationService.damageTimerAlerts);
+        }
+      }
+    }
```
Comment (mrshu) [inline lib/models/game.dart:458]: **[medium] maintainability** — Local notifications are silently gated on `VibrationService` flags.

Why: this method reads `vibrationService.gameTimerEnabled` / `damageTimerEnabled` / `gameTimerAlerts` to decide whether and what to schedule, so a class and settings section named "Vibration" also governs notifications, and a user can't enable notifications without vibration (or vice versa). A hidden cross-feature dependency the next maintainer of the settings UI will trip over.

Suggested fix: either rename `VibrationService` to a feature-neutral name (e.g. `TimerAlertService`) and relabel the settings section to say it controls both, or split out independent notification-enable flags. At minimum, document on the service and this method that the flags govern both alert mechanisms. A rename + doc is behavior-preserving.

**Review by mrshu — COMMENTED** [review]
# review-anvil report

**Target:** PR #7 "release/0.10.2" (`dev` → `main`, 56 files, +2619/-660)
**Rounds:** 1
**Mix per round:** 2 codex-exec + 1 claude-exec
**Focus:** correctness, maintainability, simplicity, production blast-radius
**Commit mode:** none (review-only)
**Auto-fix policy:** min severity = medium, allow_new_deps = false
**Review decision:** COMMENT — two `high` in-scope findings (background-timer halftime state corruption; default-enabled notifications never request OS permission) should be addressed before this release ships.

## Scope
Release 0.10.2 bundles iOS BLE support, configurable timer vibration alerts, background/local notifications for timers, editable robot labels, screen wakelock, per-module + multi-module preset saving, an MQTT refactor, the "timers continue in the background" bugfix, and a new GitHub Actions CI/CD pipeline. The two headline correctness risks both live in the new background-timer + notification path in `lib/models/game.dart` and the notification/vibration services.

**Project invariants — all preserved.** START/STOP remain fire-and-forget (`bleSendPlayAll`/`bleSendStopAll`, `timeout:0`, no new awaits on the start/stop path); destructive UI actions still use `onDoubleTap`; the `main.dart` provider tree is unchanged and static; portrait-only is intact (now also enforced on iOS).

## Round 1 — material_findings
- Reviewers: 2 codex-exec (correctness, blast-radius) + 1 claude-exec (simplicity + maintainability); all 3 returned successfully.
- Findings (post-dedup): 0 critical, 2 high, 3 medium, 2 low, 0 nit
- Fixes applied: 0 (review-only)
- Would-apply: 5 items (the 2 high + 3 medium, ≥ min_fix_severity)
- Suggestions: 2 items (low; not applied)
- Deferred: 0

---

## Actionable findings

### 1. [high] background-timer — resume catch-up replays half-time onto robots after the match has advanced to the second half
**Where:** `lib/models/game.dart` — `didChangeAppLifecycleState` catch-up loop (~L370) interacting with `halfTimeAll()` (L456) and `_maxResumeCatchUpTicks()` (L205).
**Raised by:** codex-exec (correctness) + codex-exec (blast-radius) — consensus.

On resume, `didChangeAppLifecycleState` replays missed ticks in a **synchronous** loop (`for … && isTimeRunning) _tickTimer()`). `_maxResumeCatchUpTicks()` evaluated in the first half returns `_remainingTime + halfTimeDuration`, so a long background can drive the loop through first-half expiry, into half-time, and all the way to the `halfTime → secondHalf` transition in one tight loop. When first-half hits zero, `_tickTimer` calls `halfTimeAll()`, which runs `stopAll(true)` synchronously but then `await Future.delayed(1s)` before sending `module.halfTime()`. Because the catch-up loop has no awaits, it finishes (reaching `secondHalf`, `stopAll(force:true)`) *before* that 1-second delay elapses — so the deferred `module.halfTime()` fires against modules the game already considers stopped-and-ready-for-second-half, putting the robots/displays back into a half-time countdown.

**Reachable when:** app backgrounded mid-first-half for at least `firstHalfRemaining + halfTimeDuration` seconds — entirely plausible during a real match break.

**Suggested fix:** Guard the deferred side effect on the stage still being current: in `halfTimeAll()`, after the `await`, capture/compare `currentStage` and skip the `module.halfTime()` fan-out unless `currentStage == MatchStage.halfTime`. Apply the same stale-stage guard to other delayed fan-out methods (`gameOverAll()`). Lock it in with a lifecycle test using short period/halftime values: pause, advance wall-clock past first-half + halftime, resume, wait > 1s, and assert the game ends in `secondHalf` with modules stopped — not reverted to `halfTime`.

### 2. [high] notifications — default install never requests notification permission, so background alerts silently never fire
**Where:** `lib/screens/settings.dart` L460-490 (`onChanged` handlers), `lib/services/vibration_service.dart` L8-9/L34-36 (defaults), `lib/services/notification_service.dart` L36-52 (`initialize`).
**Raised by:** all three reviewers — consensus.

`VibrationService` defaults `_gameTimerEnabled` and `_damageTimerEnabled` to `true`. `NotificationService.requestPermission()` is invoked **only** from the two settings `SettingSwitch.onChanged` handlers, and **only** when toggled to `true` (settings.dart L467, L486). `NotificationService.initialize()` deliberately does not request permission (and sets iOS `requestAlertPermission: false`). So a fresh user who keeps the default-on switches is never prompted. When the app is then backgrounded mid-timer, `Game._scheduleBackgroundNotifications()` runs (gated on those `true` flags) and calls `zonedSchedule()`, but on Android 13+ (`POST_NOTIFICATIONS`) and iOS the runtime grant was never obtained — the OS drops every alert. The headline "phone notifications when backgrounded" feature appears enabled but does nothing out of the box.

**Suggested fix:** Request permission for the default-enabled state, not only on an off→on toggle. Options: call `NotificationService.requestPermission()` from `main()` after `initialize()` (or on first foreground) when either alert is enabled; or request it the first time `_scheduleBackgroundNotifications()` is about to schedule anything. Keep the lazy on-toggle prompt for users who turn switches off then back on, and guard with a one-shot "already requested" flag to avoid re-prompting. Add a widget/service test from empty `SharedPreferences` asserting `requestPermission` is invoked before a backgrounded default-settings timer schedules notifications.

### 3. [medium] notifications — pausing during the first half schedules no half-time break alerts
**Where:** `lib/models/game.dart` `_scheduleBackgroundNotifications()` L387-390 (firstHalf case).
**Raised by:** codex-exec (correctness).

When paused in the first half, `_scheduleBackgroundNotifications()` schedules only `scheduleGameAlerts(_remainingTime, …)`. But `_maxResumeCatchUpTicks()` treats first-half + half-time as one auto-run window, so if the app stays backgrounded past first-half expiry the half-time break is considered "run" on resume — yet no break notification was ever scheduled. The referee gets a "first half over" alert but never a "half-time over / start second half" alert while the phone is away, undermining the background-alert feature for the most operationally relevant moment.

**Suggested fix:** In the `firstHalf` case, schedule both the remaining first-half alerts and the future break alerts, offsetting the break alerts by the remaining first-half duration. **This requires distinct notification IDs** — `scheduleGameAlerts` and `scheduleBreakAlerts` currently both use `10000 + threshold` (notification_service.dart L84 & L112), so co-scheduled game and break alerts would collide and overwrite each other; give break alerts their own ID base. Add a fake-scheduler test asserting that a first-half pause with `remainingTime=60`, `halfTimeDuration=30` schedules break alerts at the correct future offsets without clobbering the game alerts.

### 4. [medium] test-coverage — the app smoke test was gutted to an empty no-op as new startup paths were added
**Where:** `test/widget_test.dart` L14-31.
**Raised by:** claude-exec (maintainability).

The previously-real smoke test (`pumpWidget(MyApp(game: Game()))` + assert the UI renders) is now entirely commented out, leaving an empty `testWidgets` body. This happens in the same PR that adds new init paths — `Game()` now registers a `WidgetsBindingObserver` and constructs four new services, and `NotificationService.initialize()` runs at startup. The new `pr-quality-gate.yml` runs `flutter test`, which now passes trivially and gives false confidence that startup is verified. (The PR description does acknowledge testing is currently deferred, so this is partly a known trade-off rather than an oversight.)

**Suggested fix:** Restore a real smoke test: `SharedPreferences.setMockInitialValues({})`, construct `Game()`, `pumpWidget(MyApp(game: game))`, and assert `find.byType(Home)` is present — exercising the new observer/service construction under the test binding. Keep it green under the existing CI `flutter test` step so the gate actually guards startup.

### 5. [medium] maintainability — local notifications are silently gated on `VibrationService` flags
**Where:** `lib/models/game.dart` `_scheduleBackgroundNotifications()` L385-413.
**Raised by:** claude-exec (maintainability).

`_scheduleBackgroundNotifications()` reads `vibrationService.gameTimerEnabled` / `damageTimerEnabled` / `gameTimerAlerts` to decide whether and what to schedule. So a class and settings section named "Vibration" silently also governs notifications, and a user cannot enable notifications without vibration (or vice versa). This is a hidden cross-feature dependency that will surprise the next maintainer who edits the settings UI.

**Suggested fix:** Either rename `VibrationService` to a feature-neutral name (e.g. `TimerAlertService`) and relabel the settings section to state it controls both vibration and notifications, or split out independent notification-enable flags. At minimum, document on the service and on `_scheduleBackgroundNotifications()` that these flags govern both alert mechanisms. A pure rename + doc change is behavior-preserving.

## Suggestions
- **[low] notifications** — `lib/services/notification_service.dart` L86: the second-half/full-time path reuses `scheduleGameAlerts`, whose 0-second message reads "Time is up! Open the app to start the next timer". At full time the match is over (`gameOverAll`), so this is misleading on the most important alert. Consider a dedicated final-period message ("Full time" / "Match over"). (consider re-running with `min_fix_severity: low` to auto-apply)
- **[low] ux** — `lib/models/game.dart` resume catch-up (~L370): the synchronous replay calls `_checkGameTimerVibration()` / `notifyAllModulesTimer()` per missed tick, so crossing several alert thresholds while backgrounded fires multiple vibrations back-to-back the instant the app returns to the foreground. Consider suppressing vibration during replay (a `_replaying` flag) and resyncing once at the end, while keeping the state catch-up and MQTT publish.

## Deferred items
- None.

## Out-of-scope follow-ups
- None surfaced that meet the auto-approval or triage bar.

## Total
- Findings would-apply: 5
- Suggestions surfaced: 2
- Findings deferred: 0
- Tuning suggestion: only round was `material_findings`; consider `rounds: 2` for additional reviewer redundancy on a release-sized diff.

---
*Generated by review-anvil (read-only). No edits or commits were made.*

**Comment by mato157** [comment]
Thanks @mrshu for running the review battery — really useful. I've pushed `32a3cf0` to `dev` addressing all 7 findings (2 high + 3 medium + 2 low). All are in the background-timer/notification path:

**High**
- **H1 — stale half-time fan-out:** the resume catch-up loop replays missed seconds synchronously, so it could advance past half-time into the second half during the 1s `await` in `halfTimeAll()`/`gameOverAll()`, then fire the deferred `module.halfTime()` against robots the game had already moved past. Fixed by guarding the post-`await` fan-out on `currentStage` still being the expected stage.
- **H2 — alerts never requested permission:** the vibration/notification switches default ON but permission was only requested on an off→on toggle, so a referee who never opened settings was never prompted and background alerts were silently dropped. Now requested once on the first match START (persisted one-shot flag), unawaited so it stays off the START path.

**Medium**
- **M3 — break-alert ID collision + missing first-half break alerts:** game and break alerts both used `10000+threshold`; break alerts now use `30000+threshold`, and a first-half pause also schedules the break alerts (offset past the remaining first half) so the "start second half" alert fires across the break.
- **M4 — gutted smoke test:** restored a real `flutter test` smoke test that builds the app and asserts `Home` renders.
- **M5 — hidden Vibration→notifications coupling:** documented that those flags govern both mechanisms and relabeled the settings section "Vibration & Notifications" (behavior-preserving, kept it as one toggle group rather than splitting).

**Low**
- Full-time 0s alert now reads "Full time — match over" instead of "open the app to start the next timer".
- Per-tick vibration is suppressed during the synchronous resume replay, so crossing several thresholds while backgrounded no longer fires a burst of buzzes on return.

`flutter analyze --fatal-infos` clean, all 21 tests pass, debug APK builds. I'll re-verify the H1/H2/M3 behaviour on-device (background across first-half + break, fresh-install permission prompt). Still keeping the merge gated on @f-wllr confirming it runs clean on his iPhone.

**Comment by mato157** [comment]
Follow-up commit `50a66c1` — this is what on-device testing earns you that a green `flutter analyze` + unit suite quietly cannot. 🙂

After putting the previous batch (`32a3cf0`) on a real phone, two things showed up:

1. **Permission prompt moved to first launch.** It was firing on the first **START** press, so the OS notification dialog popped up right at kickoff. Now it's requested once on first launch (from `_loadPrefs` instead of `startTimer`), so permissions are settled before a match instead of interrupting one.

2. **Fixed a startup ordering race.** `requestPermission()` could run before `initialize()` finished — both are kicked off near-simultaneously at launch — land on the uninitialised plugin, and *silently no-op*. End result: permission would never actually be requested, which would have defeated the whole H2 fix while still looking fine in tests. `initialize()` is now idempotent (cached future) and `requestPermission()` awaits it first, so it's order-independent.

`flutter analyze --fatal-infos` clean, 21 tests pass, debug APK built and verified on-device (first-launch prompt, H1 no half-time revert on resume, M3 break alerts, no vibration burst, full-time wording). Merge still gated on @f-wllr's iPhone sign-off.

File: lib/models/game.dart [hunk 6]
```diff
@@ -118,6 +137,25 @@ class Game with ChangeNotifier {
...
+  void _maybeRequestNotificationPermission() {
+    final prefs = _prefs;
+    if (prefs == null) return;
+    if (prefs.getBool(_notifPermissionRequestedKey) ?? false) return;
+    if (!vibrationService.gameTimerEnabled &&
+        !vibrationService.damageTimerEnabled) {
+      return;
+    }
+    prefs.setBool(_notifPermissionRequestedKey, true);
+    unawaited(NotificationService.requestPermission());
```
Comment (mrshu) [inline lib/models/game.dart:156]: **[medium, non-blocking] notifications** — The one-shot flag is persisted (L155) before `requestPermission()` is known to have run.

Why: `requestPermission()` is fired unawaited and swallows init/platform-channel failures internally with no status. A transient failure at first launch (channel hiccup before the OS dialog shows) permanently burns `notif_permission_requested`, so the referee is never re-prompted and the default-on background notifications silently can never display on that device. Low probability on a real Android 13+ device, but the flag means "we attempted" rather than "the OS prompt actually fired."

Suggested fix: have `requestPermission()` report whether a platform request was actually attempted, and persist the flag only after a real attempt. Keep one-shot behavior for an actual grant/denial, but don't burn the flag on an init/channel failure. A test that mocks `requestPermission` to throw and asserts the flag is not persisted locks it in.

(Nice work overall — every round-1 finding was fixed correctly and CI is green; this is the one residual edge.)

File: lib/services/wakelock_service.dart [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  void _applyWakelock() {
+    if (kIsWeb) return;
+    try {
+      WakelockPlus.toggle(enable: _enabled);
+    } catch (e) {
+      debugPrint('WakelockService: toggle(enable: $_enabled) failed: $e');
+    }
+  }
```
Comment (mrshu) [inline lib/services/wakelock_service.dart:40]: **[low] wakelock** — `_applyWakelock()` wraps `WakelockPlus.toggle()` in a *synchronous* try/catch but doesn't `await` it, so a genuinely async platform error from the returned Future would escape this catch.

Note: this does NOT affect the restored smoke test — CI's `flutter test` is green on this commit — so there's no test/CI risk; it's only a latent runtime error-swallowing gap. Consider awaiting inside the try (make `_applyWakelock` async), or `unawaited(WakelockPlus.toggle(...).catchError(...))`, for parity with `VibrationService._initVibrator`, which awaits its plugin call inside the try/catch. Apply the same to `WakelockPlus.disable()` in `dispose()`.

**Review by mrshu — APPROVED** [review]
# review-anvil report (round 2)

**Target:** PR #7 "release/0.10.2" (`dev` → `main`), now at HEAD `50a66c1` (author pushed fixes `82f9e96..50a66c1` after round 1)
**Rounds:** 1 (this is a second, independent run against the updated head)
**Mix:** 2 codex-exec + 1 claude-exec
**Focus:** correctness, maintainability, simplicity, production blast-radius (priority: verify the round-1 fixes)
**Commit mode:** none (review-only)
**Review decision:** APPROVE — every round-1 finding was fixed and independently verified, CI is green on this commit, and the only remaining items are one non-blocking robustness `medium` and one `low`.

## Round-1 fixes — all verified correct ✅

The author addressed all five findings and both low suggestions. I confirmed each against the code (and the green CI quality gate on `50a66c1`):

| Round-1 finding | Fix | Verdict |
|---|---|---|
| [high] half-time replay onto robots after second half | `halfTimeAll()` returns early if `currentStage != halfTime` after its 1s delay; same guard on `gameOverAll()` | ✅ Correct — the synchronous resume loop can reach `secondHalf`/`fullTime` during the delay, and the stage guard suppresses the now-stale module fan-out. Stage/time broadcasts still happen in `_tickTimer`, so no display inconsistency is left. |
| [high] default install never requests notification permission | new one-shot `_maybeRequestNotificationPermission()` from `_loadPrefs`; `initialize()` made idempotent via cached `_initFuture`; `requestPermission()` now awaits it | ✅ Correct — a fresh default install (alerts default `true`) now requests permission; init/permission ordering is robust. (See finding 1 for a residual edge.) |
| [medium] first-half pause scheduled no break alerts; ID collision | firstHalf also schedules `scheduleBreakAlerts(_remainingTime + halfTimeDuration, …)`; break alerts moved to `30000 + threshold` | ✅ Correct — offset matches `_maxResumeCatchUpTicks(firstHalf)`; ID bases (game `10000+`, damage `20000+`, break `30000+`) are now disjoint. |
| [medium] gutted smoke test | restored `MyApp`→`Home` smoke test with mock prefs + fixed surface size | ✅ Verified meaningful and passing — the "Analyze, Test & Build Check" CI job (which runs `flutter analyze --fatal-infos` + `flutter test`) is **green** on `50a66c1`. |
| [medium] notifications coupled to `VibrationService` | section renamed "Vibration & Notifications"; doc comments on the service and scheduler | ✅ Sufficient (matches the "at minimum, document + relabel" fix path). |
| [low] full-time message wording | `isFinalPeriod` flag → "Full time — match over" | ✅ Correct. |
| [low] replay fires multiple vibrations | `_replaying` flag short-circuits game + damage vibration during catch-up | ✅ Correct — reset to `false` after the loop; state/MQTT updates preserved. |

**Project invariants** — all four still preserved (START/STOP fire-and-forget untouched; the new awaits are only in notification scheduling and the pre-existing `halfTimeAll`/`gameOverAll` delays, not on the robot start/stop path; double-tap, static provider tree, portrait-only unchanged).

## Round 2 — nits_only
- Reviewers: 2 codex-exec + 1 claude-exec; all 3 returned successfully (2 returned clean `[]`).
- Findings (post-dedup, post-verification): 0 critical, 0 high, 1 medium, 1 low
- Fixes applied: 0 (review-only)
- Suggestions: 1 (low)
- Deferred: 1 (failed verification)

## Actionable findings

### 1. [medium] notifications — the one-shot permission flag is burned before the request is known to have happened
**Where:** `lib/models/game.dart` L155-156 (`_maybeRequestNotificationPermission`).
**Raised by:** codex-exec (blast-radius); confirmed from code.

`_maybeRequestNotificationPermission()` persists `notif_permission_requested = true` (L155) *before* firing `unawaited(NotificationService.requestPermission())` (L156), and `requestPermission()` swallows init/platform-channel failures internally and returns no status. So a transient failure at first launch (e.g. a platform-channel hiccup before the OS dialog is shown) permanently burns the one-shot flag — the referee is never re-prompted, and the default-on background timer notifications silently can never display on that device. Low probability on a real Android 13+ device (the dialog normally shows reliably), but the flag means "we attempted" rather than "the OS prompt actually fired."

**Suggested fix:** Have `NotificationService.requestPermission()` report whether a platform request was actually attempted (return a bool, or distinguish "no plugin / threw before prompting" from "user responded"). Persist `notif_permission_requested` only after a real attempt; keep one-shot behavior for an actual grant/denial (re-prompting every launch is undesirable and impossible on iOS), but don't burn the flag on an initialization/channel failure. A regression test on the failed-request path (mock `requestPermission` to throw) asserting the flag is *not* persisted would lock this in.

## Suggestions
- **[low] wakelock** — `lib/services/wakelock_service.dart` L33-40: `_applyWakelock()` wraps `WakelockPlus.toggle(enable: _enabled)` in a *synchronous* try/catch but does not `await` it, so a genuinely asynchronous platform error from the returned `Future` would escape the catch. This does **not** affect the smoke test — CI's `flutter test` is green on this commit — so the stated "fails CI" risk is not real; it's only a latent runtime error-swallowing gap. Consider making `_applyWakelock()` (and the `dispose()` `WakelockPlus.disable()`) `await` inside the try, or use `unawaited(... .catchError(...))`, for parity with `VibrationService._initVibrator` (which awaits inside its try/catch).

## Deferred items
- **[medium→deferred] test-passability (wakelock async error fails CI)** — *failed verification.* Reviewer B predicted the restored smoke test would fail CI because `WakelockPlus.toggle()`'s async `MissingPluginException` escapes `_applyWakelock`'s synchronous try/catch. The "Analyze, Test & Build Check" job (runs `flutter test`) is **green** on `50a66c1`, empirically disproving the test-failure consequence. The underlying ineffective-try/catch is retained as the low suggestion above.

## Out-of-scope follow-ups
- **[low] notifications copy** — the first-half threshold-0 game-alert message still reads "Time is up! Open the app to start the next timer" (`notification_service.dart` ~L99). In the backgrounded auto-break case the half-time break now runs and the break alerts fire without the referee opening the app, so the wording is slightly imprecise. Pre-existing copy (round 1 only fixed the full-time wording); cosmetic, separate change.

## Total
- Findings would-apply: 1 (the medium; if `min_fix_severity: low`, the wakelock suggestion too)
- Suggestions surfaced: 1
- Findings deferred: 1 (failed verification)
- Tuning suggestion: round converged to `nits_only` — no further rounds warranted.

---
*Generated by review-anvil (read-only). No edits or commits were made. Round 1's findings were all addressed by the author in `82f9e96..50a66c1` and verified here.*

**Comment by f-wllr** [comment]
> Still keeping the merge gated on @f-wllr confirming it runs clean on his iPhone.

From my quick clicking through the app, I don't find any immediate bugs

**Review by mrshu — APPROVED** [review]
# review-anvil report (round 3 — re-review)

**Target:** PR #7 "release/0.10.2" (`dev` → `main`), now at HEAD `5cd2d89`
**Rounds:** 1 (third independent run; previously APPROVED at `50a66c1`)
**Mix:** 1 codex-exec + 1 claude-exec (reduced from 3 — proportionate to a one-line delta)
**Focus:** correctness + version consistency, release hygiene, regression re-check
**Commit mode:** none (review-only)
**Review decision:** APPROVE — the only change since the last approval is a correct, verified one-line version-string fix; no regression, CI green, no new actionable findings.

## Delta since the last approved review (`50a66c1..5cd2d89`)
One commit, `5cd2d89` *"fix(version): update version in about section to match pubspec"* — a single line in `lib/screens/settings.dart` changing the About-section text from `Version: 0.9.8` to `Version: 0.10.2`.

**Verified correct and complete:**
- Matches `pubspec.yaml` (`version: 0.10.2+12`). ✅
- No other stale user-visible version string remains in `lib/`/`android/`/`ios/` (repo-wide grep; the only hit is the corrected literal). ✅
- Pure display text — touches no logic, provider tree, BLE/MQTT path, double-tap handlers, or orientation lock. All four project invariants re-confirmed intact by both reviewers. ✅
- CI "Analyze, Test & Build Check" (`flutter analyze --fatal-infos` + `flutter test`) and "Build Debug APK" are both **green** on `5cd2d89`. ✅

## Round 3 — clean
- Reviewers: 1 codex-exec + 1 claude-exec; both returned successfully, both `[]`.
- Findings: 0 critical, 0 high, 0 medium, 0 low, 0 nit
- Fixes applied: 0 (review-only)
- Would-apply: 0
- Suggestions: 0
- Deferred: 0

## Standing non-blocking items (round 2, unchanged — author's call)
Not re-raised; the `5cd2d89` commit does not touch them:
- [medium] `game.dart` `_maybeRequestNotificationPermission()` persists the one-shot flag before the unawaited `requestPermission()` is confirmed.
- [low] `wakelock_service.dart` `_applyWakelock()` synchronous try/catch can't catch the un-awaited `WakelockPlus.toggle()` async error (no test/CI impact; CI green).

## Out-of-scope follow-ups
- **[low / needs_triage] release hygiene** — the About-screen version is a hardcoded literal that must be hand-synced with `pubspec.yaml` on every release (exactly the drift this commit corrects — it had been left at `0.9.8` across the `0.10.x` bumps). A more robust pattern reads it at runtime via `package_info_plus` (`PackageInfo.fromPlatform().version`), which can't go stale. Same class as the harmless pre-existing `flutterVersionName = "0.9.2"` Gradle fallback. Separate, non-blocking change.

## Total
- Findings would-apply: 0
- Suggestions surfaced: 0
- Findings deferred: 0
- Converged: clean — no further rounds warranted.

---
*Generated by review-anvil (read-only). No edits or commits were made.*

**Comment by mrshu** [comment]
@f-wllr @mato157 let's get this in!
